# Numerical Methods week 2

### Learning outcomes:
- Recall methods of solution of inhomogeneous systems of linear equations.
- Elimination methods, Gauss elimination, Gauss-Jordan elimination .
- Implement and use Gauss-Jordan Elimination to solve systems of equations.
  
### Reading:

- Introduction to Part 3 and chapter 9 of Chapra and Canale.

## Solving a system of Equations

Suppose we want to solve a system of equations $A x = b$

Where $A$ is a matrix of coefficients of our vector of unknowns, $x$ and $b$ is a vector of constants.

Explicitly this can be written (for a set of 4 equations with 4 unknowns) as:

$$ 
\left( \begin{array}{cccc}
a_{00} & a_{01} & a_{02} & a_{03}\\ 
a_{10} & a_{11} & a_{12} & a_{13}\\ 
a_{20} & a_{21} & a_{22} & a_{23}\\ 
a_{30} & a_{31} & a_{32} & a_{33}\\ 
\end{array} \right)
\left( \begin{array}{c}
x_0\\
x_1\\
x_2\\
x_3\\
\end{array} \right)
=
\left( \begin{array}{c}
b_0\\
b_1\\
b_2\\
b_3\\
\end{array} \right)
$$

### Augmented Matrix

<p class='fragment'> It can be useful to rewrite this as an <a href='https://en.wikipedia.org/wiki/Augmented_matrix'>augmented matrix</a><br><br> 
   \[ \left( \begin{array}{cccc|c}
   a_{00} & a_{01} & a_{02} & a_{03} & b_0\\ 
   a_{10} & a_{11} & a_{12} & a_{13} & b_1\\ 
   a_{20} & a_{21} & a_{22} & a_{23} & b_2\\ 
   a_{30} & a_{31} & a_{32} & a_{33} & b_3\\ 
   \end{array} \right)\]
 </p>

   <h2 id="Gaussian Elimination">Gaussian Elimination<a class="anchor-link" href="#Gaussian Elimination">&#182;</a></h2>    
   <h3 id="Triangularization">Triangularization<a class="anchor-link" href="#Triangularization">&#182;</a></h3>      
   <p>We reduce the augmented matrix to <a href='https://en.wikipedia.org/wiki/Row_echelon_form'>row echelon form</a>. Let us take an initial augmented matrix as: 
     \[ \left( \begin{array}{cccc|c}
     2 & 2 & 4 & -2 & 10\\ 
     1 & 3 & 2 & 4 & 17\\ 
     3 & 1 & 3 & 1 & 18\\ 
     1 & 3 & 4 & 2 & 27\\ 
     \end{array} \right) \]
   </p>
   <p class=fragment>
     pivoting around row 0, we remove all entries below the diagonal entry in column 0,
     \[ \left( \begin{array}{cccc|c}
     2 & 2 & 4 & -2 & 10\\ 
     0 & 2 & 0 & 5 & 12\\ 
     0 & -2 & -3 & 4 & 3\\ 
     0 & 2 & 2 & 3 & 22\\ 
     \end{array} \right) \]
   </p>

   <p>
     Matrix after pivoting around row 0	
     \[ \left( \begin{array}{cccc|c}
     2 & 2 & 4 & -2 & 10\\ 
     0 & 2 & 0 & 5 & 12\\ 
     0 & -2 & -3 & 4 & 3\\ 
     0 & 2 & 2 & 3 & 22\\ 
     \end{array} \right) \]
   </p><p class=fragment>
     Then pivoting around row 1 we remove elements below the diagonal in column 1,
     \[ \left( \begin{array}{cccc|c}
     2 & 2 & 4 & -2 & 10\\ 
     0 & 2 & 0 & 5 & 12\\ 
     0 & 0 & -3 & 9 & 15\\ 
     0 & 0 & 2 & -2 & 10\\ 
     \end{array} \right) \]
   </p><p class=fragment>	  
     pivoting around row 2
     \[ \left( \begin{array}{cccc|c}
     2 & 2 & 4 & -2 & 10\\ 
     0 & 2 & 0 & 5 & 12\\ 
     0 & 0 & -3 & 9 & 15\\ 
     0 & 0 & 0 & 4 & 20\\ 
     \end{array} \right) \]
   </p>

  <h2 id="Gaussian Elimination">Gaussian Elimination<a class="anchor-link" href="#Gaussian Elimination">&#182;</a></h2>    
  <h3 id="Back substitution">Back substitution<a class="anchor-link" href="#Back substitution">&#182;</a></h3>
  <p> From the triangularized augmented matrix we can then solve for $x$
   \[ \left( \begin{array}{cccc|c}
   2 & 2 & 4 & -2 & 10\\ 
   0 & 2 & 0 & 5 & 12\\ 
   0 & 0 & -3 & 9 & 15\\ 
   0 & 0 & 0 & 4 & 20\\ 
   \end{array} \right) \]
 </p>
 <p class=fragment>Starting at row 3, only the coefficient of $x_3$ is non-zero. Converting the augmented notation to real equations using the last row we have
   $$
   0 \cdot x_0 + 0 \cdot x_1 + 0 \cdot x_2 + 4 \cdot x_3 = 20 \implies x_3 = 5
   $$
 </p>
 <p class=fragment>Then we can work up the rows
   \[\begin{align*}
   0 \cdot x_0 + 0 \cdot x_1 + -3 \cdot x_2 + 9 \cdot x_3 & = 15 \\
   0 \cdot x_0 + 0 \cdot x_1 + -3 \cdot x_2 + 9 \cdot 5 & = 15 \implies x_2 = 10	  
   \end{align*}\]
 </p>

  <p class=fragment>Now row 2
   \[\begin{align*}
   0 \cdot x_0 + 2 \cdot x_1 + 0 \cdot x_2 + 5 \cdot x_3 & = 12 \\
   0 \cdot x_0 + 2 \cdot x_1 + 0 \cdot 10 + 9 \cdot 4 & = 12 \implies x_1 = -6.5
   \end{align*}\]
 </p>
 <p class=fragment>And finally
   \[\begin{align*}
   2 \cdot x_0 + 2 \cdot -6.5 + 4 \cdot 10 + -2 \cdot 4 & = 10 \implies x_0 = -3.5
   \end{align*}\]
 </p>

### Python Gauss Elimination

Ideally I would like everyone to be able to use C++ or python. Here is a python code.

You can run python code using spyder, a python shell or jupyter notebooks. All are available on the boxes in INB 2305 and are freely available online.  


In [22]:
import numpy as np # import our numpy library and call it np for short

N = 3 # size of the problem
a = np.random.uniform(1,10,(N,N+1)) # generate a N x(N+1) matrix 
x = np.zeros(3) # create a solution vector of zeros ready to be filled
print(a)

[[1.85973119 9.09563155 8.93534151 8.68534022]
 [9.55680619 6.52389492 1.51861023 8.86640528]
 [1.72905822 2.90730446 6.61054075 7.08207885]]


In [24]:
# eliminating - we are going to modify a as we do it
for i in range(N-1): # go across the columns 1 at a time
    for j in range(i+1,N,1): # now we go down the column below the diagonal
        coeff = a[j,i]/a[i,i] # find ratio of diagonal element to value in this row
        for k in range(i,N+1,1): # no move along the row
            a[j,k] -= a[i,k]*coeff # subtract the right amoung of the row on the diagonal 
print(a)

[[  1.85973119   9.09563155   8.93534151   8.68534022]
 [  0.         -40.21683221 -44.39841678 -35.76591245]
 [  0.           0.           4.42925003   3.94208684]]


In [25]:
# now back substitute
for i in range(N-1,-1,-1): # start at the bottom and work up
    x[i] = a[i,N+1-1]
    for j in range(i+1, N,1):
        x[i] -= a[i,j]*x[j]
    x[i] /= a[i,i]
print(x)

[ 0.84997173 -0.09322522  0.89001226]


To make things more usable we could put the Guassian Elimination code into a function:

In [27]:
import numpy as np

def gauss_elim(a, N):
    """solves a set of linear equations
    N must be the number of rows (dimension of solution). 
    Warning: On exit the original matrix is changed to row echelon form.
    """
    x = np.zeros(N);
    # eliminating
    for i in range(N-1):
        for j in range(i+1,N,1):
            coeff = a[j,i]/a[i,i]
            for k in range(i,N+1,1):
                a[j,k] -= a[i,k]*coeff
    # now back substitute
    for i in range(N-1,-1,-1):
        x[i] = a[i,N+1-1]
        for j in range(i+1, N,1):
            x[i] -= a[i,j]*x[j]
        x[i] /= a[i,i]
    return x

# build a random 2d 3x4 augmented matrix
x = gauss_elim(a,N)      
print(x)

[ 0.84997173 -0.09322522  0.89001226]


  <h2 id="Gauss Elimination">Gauss Elimination<a class="anchor-link" href="#Gauss Elimination">&#182;</a></h2>
  <p>Use the code to find the solutions of the following systems</p>

  <div style="background-color:Lavender; margin-left: 20px; margin-right: 20px; padding-bottom: 8px; padding-left:
  8px; padding-right: 8px; padding-top: 8px; border-radius: 25px;">
  <p>
   \[
   \begin{align*}
   3x_0  +  4x_1 - 7x_2 & = 23\\
   7x_0  -   x_1 + 2x_2 & = 14\\
   x_0  + 10x_1 - 2x_2 & = 33\\
   \end{align*}
   \]
 </p>
 <p>Can you find the solutions to this system of equations? Why not?
   \[
   \begin{align*}
   1x_0 + 2x_1 + 3x_2 & = 1\\
   4x_0 + 5x_1 + 6x_2 & = 2\\
   7x_0 + 8x_1 + 9x_2 & = 3\\
   \end{align*}
   \]<br><br>
 </p>
</div>

## Guass-Jordan elimination

<p>Guass-Jordan elimination is very similar to Gauss elimination. Instead of triangularization, we make a completely diagonal matrix. Or more exactly we reduce the augmented matrix to $\textbf{reduced}$ row echelon form.
     Initial matrix is: 
     \[ \left( \begin{array}{cccc|c}
     2 & 2 & 4 & -2 & 10\\ 
     1 & 3 & 2 & 4 & 17\\ 
     3 & 1 & 3 & 1 & 18\\ 
     1 & 3 & 4 & 2 & 27\\ 
     \end{array} \right) \]
     as before
   </p><p class='fragment'> The first step is the same:	  
     pivoting around row 0
     \[ \left( \begin{array}{cccc|c}
     2 & 2 & 4 & -2 & 10\\ 
     0 & 2 & 0 & 5 & 12\\ 
     0 & -2 & -3 & 4 & 3\\ 
     0 & 2 & 2 & 3 & 22\\ 
     \end{array} \right) \]
   </p><p>
   </p>

The first step is the same:	  
pivoting around row 0
       
$$ \left( \begin{array}{cccc|c}
2 & 2 & 4 & -2 & 10\\ 
0 & 2 & 0 & 5 & 12\\ 
0 & -2 & -3 & 4 & 3\\ 
0 & 2 & 2 & 3 & 22\\ 
\end{array} \right) $$


But now, pivoting around row 1, we remove entries above $\textbf{and}$ below the diagonal of column 1

$$
\left( \begin{array}{cccc|c}
       2 & 0 & 4 & -7 & -2\\ 
       0 & 2 & 0 & 5 & 12\\ 
       0 & 0 & -3 & 9 & 15\\ 
       0 & 0 & 2 & -2 & 10\\ 
       \end{array} \right)
$$

This continues
       pivoting around row 2
       $$ \left( \begin{array}{cccc|c}
       2 & 0 & 0 & 5 & 18\\ 
       0 & 2 & 0 & 5 & 12\\ 
       0 & 0 & -3 & 9 & 15\\ 
       0 & 0 & 0 & 4 & 20\\ 
       \end{array} \right) $$
     </p>

<p>And finally, pivoting around row 3

$$ \left( \begin{array}{cccc|c}
2 & 0 & 0 & 0 & -7\\ 
0 & 2 & 0 & 0 & -13\\ 
0 & 0 & -3 & 0 & -30\\ 
0 & 0 & 0 & 4 & 20\\ 
\end{array} \right) $$

</p>
<p>We can then divide each row by the final coefficients to get: 
       $$ \left( \begin{array}{cccc|c}
       1 & 0 & 0 & 0 & -3.5\\ 
       0 & 1 & 0 & 0 & -6.5\\ 
       -0 & -0 & 1 & -0 & 10\\ 
       0 & 0 & 0 & 1 & 5\\ 
       \end{array} \right) $$
       And we can just read off the solutions, $x$.
     </p>

<h2 id="Gauss-Jordan Elimination for Matrix Inversion">Gauss-Jordan Elimination for Matrix Inversion<a class="anchor-link" href="#Gauss-Jordan Elimination for Matrix Inversion">&#182;</a></h2>
   
<p>We can solve the equation $AB = I$ using exactly the same method:
 \[ \left( \begin{array}{ccc|ccc}
 2 & 1 & 1 & 1 & 0 & 0\\ 
 1 & 0 & -1 & 0 & 1 & 0\\ 
 2 & -1 & 2 & 0 & 0 & 1\\ 
 \end{array} \right) \]

pivoting around row 0

$$ \left( \begin{array}{ccc|ccc}
 2 & 1 & 1 & 1 & 0 & 0\\ 
 0 & -0.5 & -1.5 & -0.5 & 1 & 0\\ 
 0 & -2 & 1 & -1 & 0 & 1\\ 
 \end{array} \right) $$

 pivoting around row 1

$$ \left( \begin{array}{ccc|ccc}
 2 & 0 & -2 & 0 & 2 & 0\\ 
 0 & -0.5 & -1.5 & -0.5 & 1 & 0\\ 
 0 & 0 & 7 & 1 & -4 & 1\\ 
 \end{array} \right) $$
</p>

   <p>
     pivoting around row 2
     \[ \left( \begin{array}{ccc|ccc}
     2 & 0 & 0 & 0.285714 & 0.857143 & 0.285714\\ 
     0 & -0.5 & 0 & -0.285714 & 0.142857 & 0.214286\\ 
     0 & 0 & 7 & 1 & -4 & 1\\ 
     \end{array} \right) \]

 Scaling the rows, the final matrix is:

 $$ \left( \begin{array}{ccc|ccc}
 1 & 0 & 0 & 0.142857 & 0.428571 & 0.142857\\ 
 -0 & 1 & -0 & 0.571429 & -0.285714 & -0.428571\\ 
 0 & 0 & 1 & 0.142857 & -0.571429 & 0.142857\\ 
 \end{array} \right)$$
   </p>
   <p> The RHS of the augmented matrix is $A^{-1}$
   </p>

  <h2 id="Gauss-Jordan Elimination">Gauss-Jordan Elimination<a class="anchor-link" href="#Gauss-Jordan Elimination">&#182;</a></h2>
  <div style="background-color:Lavender; margin-left: 20px; margin-right: 20px; padding-bottom: 8px; padding-left:
  8px; padding-right: 8px; padding-top: 8px; border-radius: 25px;">
    
    
  <p>Alter the code for Gauss elimination to instead perform Gauss-Jordan elimination.<br><br>
   - I suggest copying your file for now, renaming rather than altering the previous code directly.
   - You should change the second loop so that it goes over all rows.<br>
   - You should add an `if` statement to skip the row with the same index as the column you are working on.<br>
   - When the matrix is diagonal, divide each row by the value of the remaining diagonal element to get the identity matrix and $x$.<br>
   - Test regularly as you make the alterations.
 </p>
  <p> Solve the same matrix problems as before and check the result is the same.</p>

<h2 id="Gauss-Jordan Elimination">Gauss-Jordan Matrix inversion<a class="anchor-link" href="#Gauss-Jordan Elimination">&#182;</a></h2>
<div style="background-color:Lavender; margin-left: 20px; margin-right: 20px; padding-bottom: 8px; padding-left:
8px; padding-right: 8px; padding-top: 8px; border-radius: 25px;">
<p>Make a new Gauss-Jordan routine that can calculate the inverse of a matrix.<br><br>
 - You need to extend the number of columns in the augmented matrix<br>
 - You need to extend range of the loops that go over the columns<br><br>
<p> Find the inverse of the matrix
 \[ \left( \begin{array}{cccc}
 2 & 2 & 4 & -2 \\ 
 1 & 3 & 2 & 4 \\ 
 3 & 1 & 3 & 1\\ 
 1 & 3 & 4 & 2 \\ 
 \end{array} \right) \]
 <br>
 Check your solution is correct by multiplying the original and inverse matrices.
</p>	    
</div>

  <h2 id="Test yourself">Test yourself<a class="anchor-link" href="#Test yourself">&#182;</a></h2>    
  <p>Try the test in the assessments tab of Bb. It is based on last week's material. The difficulty will probably be working with the matrices. Try to get a feel for the indices and accessing individual elements.</p>
  </div>
  
<h2 id="Summary and Further Reading">Summary and Further Reading<a class="anchor-link" href="#Summary and Further Reading">&#182;</a></h2>
  <p>You should be reading additional material to provide a solid background to what we do in class</p>
  <p>All the textbooks in the book list on Bb contain sections on solving linear equations. I suggest Chapter 9 of Chapra and Canale for starters.</p>
  <h3>Homework</h3>
  <p>Before next week read about extra steps that can be performed to improve elimination methods.</p>
  <p>Read about LU decomposition of square matrices, Chapter 10 of Chapra and Canale.</p>